sycl/test-e2e/Annotated_usm/annotated_usm.cpp

```diff
@@ -0,0 +1,173 @@
+#define TEST_ALIGN(f, align, args...)                                          \
+  {                                                                            \
+    auto ap = f(args);                                                         \
+    assert(ap.get() != nullptr && ((uintptr_t)ap.get() % align) == 0);         \
+  }
+
+// Single test instance consisting of (i) calling malloc function `f` (ii)
+// checking return is nullptr (iii) calling free
+#define TEST_ALIGN_NULL_EXPECTED(f, args...)                                   \
+  {                                                                            \
+    auto ap = f(args);                                                         \
+    assert(ap.get() == nullptr);                                               \
+    free(ap, q);                                                               \
+  }
+
+constexpr int N = 10;
+
+template <typename T> void testUsmKind(sycl::queue &q) {
+  const sycl::context &Ctx = q.get_context();
+  auto dev = q.get_device();
+
+  TEST_USM_KIND(malloc_annotated<T>, alloc::device, N, q, alloc::device);
+  TEST_USM_KIND(malloc_annotated, alloc::device, N, q, alloc::device);
+  TEST_USM_KIND(aligned_alloc_annotated<T>, alloc::host, 1, N, q, alloc::host);
+  TEST_USM_KIND(aligned_alloc_annotated, alloc::host, 1, N, q, alloc::host);
+
+  TEST_USM_KIND(malloc_device_annotated, alloc::device, N, q);
+  TEST_USM_KIND(malloc_device_annotated<T>, alloc::device, N, q);
+  TEST_USM_KIND(aligned_alloc_device_annotated, alloc::device, 0, N, q);
+  TEST_USM_KIND(aligned_alloc_device_annotated<T>, alloc::device, 0, N, q);
+
+  TEST_USM_KIND(malloc_host_annotated, alloc::host, N, q);
+  TEST_USM_KIND((malloc_host_annotated<T>), alloc::host, N, q);
+  TEST_USM_KIND(aligned_alloc_host_annotated, alloc::host, 0, N, q);
+  TEST_USM_KIND((aligned_alloc_host_annotated<T>), alloc::host, 0, N, q);
+
+  if (dev.has(sycl::aspect::usm_shared_allocations)) {
+    TEST_USM_KIND(malloc_annotated, alloc::shared, N, q,
+                  properties{usm_kind_shared});
+    TEST_USM_KIND(malloc_annotated<T>, alloc::shared, N, q,
+                  properties{usm_kind_shared});
+
+    TEST_USM_KIND(malloc_shared_annotated, alloc::shared, N, q);
+    TEST_USM_KIND((malloc_shared_annotated<T>), alloc::shared, N, q);
+    TEST_USM_KIND(aligned_alloc_shared_annotated, alloc::shared, 0, N, q);
+    TEST_USM_KIND((aligned_alloc_shared_annotated<T>), alloc::shared, 0, N, q);
+  }
+}
+
+template <typename T> void testAlign(sycl::queue &q, unsigned align) {
+  const sycl::context &Ctx = q.get_context();
+  auto dev = q.get_device();
+
+  assert(align > 0 || (align & (align - 1)) == 0);
+
+  // Case: malloc_xxx with no alignment property, the alignment is sizeof(T)
+  TEST_ALIGN((malloc_device_annotated<T>), sizeof(T), N, q);
+  TEST_ALIGN((malloc_host_annotated<T>), sizeof(T), N, q);
+  if (dev.has(sycl::aspect::usm_shared_allocations)) {
```

Comment:
Can you remove this device check? We want to test shared USM. If the device doesn't support it, the device shouldn't be used to run this test.

Reply:
There is possibility to use `// REQUIRES: aspect-usm_shared-allocations` to do that, I believe.

Reply:
> Removing dev.has(sycl::aspect::usm_shared_allocations) will lose the test coverage for shared allocation

Did you mean for "device allocation"?

Reply:
If we remove the whole block of 
```
if (dev.has(sycl::aspect::usm_shared_allocations)) {
  // call shared alloc functions
}
``` 
then we lose the tests for shared allocation on the platforms that support shared allocation

Reply:
Update: split the test into two files: `annotated_usm_align.cpp` (for device/host) and `annotated_usm_align_shared.cpp` (for shared). And add `// REQUIRES: aspect-usm_shared_allocations` to the latter